### PR TITLE
ON-47268 # Updated freshdeskService to support choices from freshdesk

### PR DESCRIPTION
**Pull request by jdawg093**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [x] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

### CODE REVIEW

File: src/freshdeskService.ts [hunk 1]
```diff
@@ -163,3 +169,56 @@ export function generateFreshdeskTicketFieldDefinitions(
...
+/**
+ * This internal function is used to decide if `options` or `choices` should be
+ * used. If there is no `options` parameter passed in, it will use the `choices`
+ * configuration to create the options required for the options required.
+ *
+ * @param options
+ * @param choices
+ * @returns
+ */
+function transformFreshdeskChoicesIntoOptions(
```
Comment (mymattcarroll) [inline src/freshdeskService.ts:182]: Need to export this so it can used in the API.

File: src/freshdeskService.ts [hunk 2]
```diff
@@ -163,3 +169,56 @@ export function generateFreshdeskTicketFieldDefinitions(
...
+/**
+ * This function is used to decide if `options` or `choices` should be used. If
+ * there is no `options` parameter passed in, it will use the `choices`
+ * configuration to create the options required for the options required.
+ *
+ * @param options
+ * @param choices
+ * @returns
+ */
+export function transformFreshdeskChoicesIntoOptions(
```
Comment (mymattcarroll) [inline src/freshdeskService.ts:182]: Prob makes more sense to make this function take in `FreshdeskTypes.FreshdeskField[]` and return `FreshdeskTypes.FreshdeskField[]` and name the function `setFreshdeskFieldOptions()`. Makes it a bit clearer what is doing and how to call it.

File: tests/freshdesk.test.ts [hunk 3]
```diff
@@ -240,4 +240,140 @@ describe('generateFreshdeskTicketFieldDefinitions', () => {
...
+        id: '3',
+        isRequired: true,
+        type: 'CHOICE_SINGLE',
+      },
+    ])
+  })
+
+  it('should map DEPENDENT_FIELD field correctly when choices is used', () => {
+    const fields: FreshdeskTypes.FreshdeskField[] = [
+      //@ts-expect-error Not filling in all properties
```
Comment (mymattcarroll) [inline tests/freshdesk.test.ts:280]: Can we put our freshdesk fields here instead of just some fields.

File: tests/freshdesk.test.ts [hunk 4]
```diff
@@ -240,4 +240,140 @@ describe('generateFreshdeskTicketFieldDefinitions', () => {
...
+            QLD: ['Brisbane', 'Gold Coast'],
+          },
+          USA: {
+            Texas: ['Austin', 'Dallas'],
+            Kansas: ['Kansas City', 'Manhattan'],
+          },
+        },
+      },
+    ]
+    expect(generateFreshdeskTicketFieldDefinitions(fields)).toEqual([
```
Comment (mymattcarroll) [inline tests/freshdesk.test.ts:299]: Prob better to use a snapshot.

File: tests/freshdesk.test.ts (hunk 3 above)
Comment (jdawg093) [reply tests/freshdesk.test.ts:280]: To clarify: are you saying to put all of our freshdesk fields into this one test?

File: tests/freshdesk.test.ts (hunk 4 above)
Comment (mymattcarroll) [reply tests/freshdesk.test.ts:299]: yes

File: tests/freshdesk.test.ts (hunk 3 above)
Comment (mymattcarroll) [reply tests/freshdesk.test.ts:280]: Yes

Comment (jdawg093) [reply tests/freshdesk.test.ts:280]: It seems a bit overkill to have all 16 different fields here for the test. I think it would be better to have the test include a mixture of fields options + choices and one without choices to ensure they all come out correct. Is there a reason to have all the freshdesk fields?

Comment (mymattcarroll) [reply tests/freshdesk.test.ts:280]: Pass them all in, check snapshot value, regressions are now impossible.

Could pass in individual ones if you want but a lot more work for the same result and more maintenance.

Comment (mymattcarroll) [reply tests/freshdesk.test.ts:280]: In the same test, run the function again with the result as the input and assert the input is the same as the output. That will cover all scenarios.

Comment (jdawg093) [reply tests/freshdesk.test.ts:280]: Seems like this scenario would be better as a new test as opposed to including in this test. Then, we have a test that ensures the functionality of the mapping and we have a test that ensures the fields are transforming correctly.

File: tests/freshdesk.test.ts [hunk 5]
```diff
@@ -240,4 +244,73 @@ describe('generateFreshdeskTicketFieldDefinitions', () => {
...
+})
+
+describe('setFreshdeskFieldOptions', () => {
+  it('should transform all of the fields correctly', () => {
+    const freshdeskFields = setFreshdeskFieldOptions(
+      freshdeskFieldFixture as FreshdeskTypes.FreshdeskField[],
+    )
+    expect(freshdeskFields).toMatchSnapshot()
+
+    expect(setFreshdeskFieldOptions(freshdeskFields)).toMatchSnapshot()
```
Comment (mymattcarroll) [inline tests/freshdesk.test.ts:314]:
```suggestion
    expect(setFreshdeskFieldOptions(freshdeskFields)).toEqual(freshdeskFields)
```

The value here is knowing that it has not been changed by another iteration.

Comment (jdawg093) [reply tests/freshdesk.test.ts:314]: yep, just remembered then before checking this, completely slipped my mind. classic.
